pyshepseg/tiling.py

```diff
@@ -738,6 +743,212 @@ def crossesMidline(overlap, segLoc, orientation):
     return ((minN < mid) & (maxN >= mid))
 
 
+def calcPerSegmentStatsTiled(imgfile, imgbandnum, segfile, maxSegId, 
+            statsSelection, chunkSize=DFLT_CHUNKSIZE):
+    """
+    Calculate selected per-segment statistics for the given band 
+    of the imgfile, against the given segment raster file. 
+    Calculated statistics are written to the imgfile raster 
+    attribute table (RAT), so this file format must support RATs. 
+    
+    Calculations are carried out in a memory-efficient way, allowing 
+    very large rasters to be processed. Raster data is handled in 
+    small tiles, attribute table is handled in fixed-size chunks. 
+    
+    The statsSelection parameter is a list of tuples, one for each
+    statistics to be included. Each tuple is either 2 or 3 elements,
+        (columnName, statName) or (columnName, statName, parameter)
+    The 3-element form is used for any statistic which requires
+    a parameter, which currently is only the percentile. 
+    
+    The columnName is a string, used to name the column in the 
+    output RAT. 
+    The statName is a string used to identify which statistic 
+    is to be calculated. Available options are:
+        'min', 'max', 'mean', 'stddev', 'mode', 'percentile'.
+    The 'percentile' statistic requires the 3-element form, with 
+    the 3rd element being the percentile to be calculated. 
+    
+    For example
+        [('Band1_Mean', 'mean'),
+         ('Band1_stdDev', 'stddev'),
+         ('Band1_LQ', 'percentile', 25),
+         ('Band1_UQ', 'percentile', 75)
+        ]
```

Comment:
Should there be usage in there also? ie `gdal.GFU_Generic`, `gdal.GFU_Red` etc?

Reply:
No, because this code does not do the actual colour columns. Should it? I suspect not, as those colour columns will be derived just from the stats columns, rather than from the imagery itself. I _think_ so, anyway....

Reply:
Umm I was thinking that in most cases it would be this code that creates the colour columns (there would have to be some mechanism to clip 0-255). Or are you thinking there would be a separate pass over the RAT do to this?

Reply:
Yes, I was thinking a separate pass, just over the RAT. There is a whole extra calculation to do, involving how to stretch, which is done from other columns. If we try to build that into this code, it would require some further special statistic types to define how that stretch should be done (e.g. a 2 std dev stretch, or whatever), as well as the clip to 0-255. However, it should not need to involve the imagery, once the basic stats are done. So, it seemed to me that this was a separate task.

Reply:
ok fair enough. Maybe we should depend on `rios.ratapplier` here rather than have to write the whole thing again...

Reply:
Yes

Reply:
Upon reflection, I don't think this really works with ratapplier, either. To use, say, a column of per-segment mean values to make a column of Red colour values, one needs the whole column all at once to get the lower and upper bounds (by whatever method) in order to stretch. So, it doesn't really lend itself to chunk-by-chunk work. Dunno, we will need to ponder....

Reply:
What about http://www.rioshome.org/en/latest/rios_fileinfo.html#rios.fileinfo.RatStats ?

Reply:
Yes, true enough. That involves a pass through the column to gather relevant statistics, and then another pass to calculate the stretched colour column. The current rios RatStats implementation would not allow stretches like the one we used in the current pyshepseg/utils.py, which does 5th and 95th percentile stretch, so we would need to write something specific. 

Dunno..... I am feeling like it would be easier just to read the whole single column and stretch it all at once. However, I am not committed either way.

Reply:
Ah yes, that is a bit of a trap isn't it. Something to add to RIOS in future? :smile: 

OK let's read the whole column. We are going to have to do that for calculating the stats from the histogram anyway. And if we are only going to be reading one column at a time into memory it won't be too bad. And maybe we don't need to worry about the Raspberry PI case ATM...

Reply:
:-)